### PR TITLE
finalize: separate the ALLFUNC_MUTEX from all other mutexes

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

Current development is adding dynamically allocated mutexes so that de-allocation has to be before we call finalize_memory_tracing. So we need separate the `ALLFUNC` mutex from the rest of the run-time mutexes just like the way we do in `init`.

Apply the similar thread cs refactor to finalize as we did in init, so
that the critical section now protects the entire finalize function.

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Passes tests (included warning check)
* [x] Confirm whitespace/style checkers are happy (or has a good reason for being bad)
* [x] Commits are self-contained and do not do two things at once
* [ ] Remove xfail from the test suite when fixing a test
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Add comments such that someone without knowledge of the code could understand
* [ ] Add Devel Docs in the `doc/` directory for any new code design

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/warnings

**Comment by hzhou** [comment]
test:mpich/warnings

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most
test:mpich/warnings

File: src/mpi/init/init_thread_cs.c [hunk 1]
```diff
@@ -194,7 +207,7 @@ void MPIR_Thread_CS_Finalize(void)
 
 #else
 /* not MPICH_IS_THREADED, empty stubs */
-void MPII_init_thread_and_enter_cs(void)
+void MPII_init_thread_and_enter_cs(int thread_required)
```
Comment (raffenet) [inline src/mpi/init/init_thread_cs.c]: nit: this fixup is unrelated to the rest of the commit.

**Comment by rzambre** [comment]
The first commit of this PR, unfortunately, breaks the per-vci critical section during `MPI_Finalize` since the vci lock is destroyed as soon as we enter `MPI_Finalize`. There is an Allreduce during OFI's finalize routine which requires the vci lock for the collective's pt2pt operations.

**Comment by hzhou** [comment]
@rzambre I'll make another patch to fix it, by moving the destroy after MPID_Finalize. They insist on using `MPIR_Allreduce` to serve as a barrier ... :(  Ideally, we should call `MPI_Barrier` at the beginning of `MPI_Finalize`, but `ofi` need some hack right now to ensure remote completion of the barrier itself, so the "barrier" lives deep inside ofi for now.
